Lon-lat variable sediment porosity

jmaerz opened the pull request:
This feature branch enables to read a sediment porosity file as input for the sediment compartment of iHAMOCC. iHAMOCC then considers the spatially variable porosity through the scaling of the sediment pore water diffusion coefficients and in sediment processes. This feature can thus far only be switched on via the namelist switch `l_sed_por` for expert users (current default is `.false.`). 

The reading of a sediment porosity file is thus far untested (and will produce results different from CMIP6), while the current default version with `l_sed_por=.false.` produces bit identical results compared to the former master branch. I am working on the input file for this feature, but wanted to give you already the possibility to review the code changes.

jmaerz committed:
ADD oafx to meson

jmaerz committed:
Declaring and allocating 3-D fields for porsol, porwah, porwat

jmaerz committed:
Introducing 2D field for total solid sediment volume: solfu

jmaerz committed:
Initializing 2D and 3D fields for sediment

jmaerz committed:
deleted real solfu, since it became an array

jmaerz committed:
exchange k by i,j,k for porwat,porsol,porwah

jmaerz committed:
make sediment diffusion coeff spatially (lon,lat,z) dependent of porosity

jmaerz committed:
merge with fixed oafx branch

jmaerz committed:
fix import of zcoefsu

jmaerz committed:
fields to 3D

jmaerz committed:
to 3D and account for lon-lat inhomogeneous porosity

jmaerz committed:
to 2D / 3D

jmaerz committed:
1D -> 3D

jmaerz committed:
Initialize parameters for dissolution/remin already in bodensed - not hard-coded in powach

jmaerz committed:
some clean-up

jmaerz committed:
Introduce potential for reading and application of lon-lat variable sediment porosity

jmaerz commented:
Ok, I just realize that the commit comes with all changes done in master over the past days (as compared to this development branch). Maybe it would be good to first merge the master into the development branch?

jmaerz committed:
merge with master - fixed oafx

TomasTorsvik commented:
> Ok, I just realize that the commit comes with all changes done in master over the past days (as compared to this development branch). Maybe it would be good to first merge the master into the development branch?

Yes, probably that is a good idea, although this will create an additional merge commit on the development branch. Don't know if there is any better alternative though.

UPDATE: I created a PR for merging `master` to `feature-hamocc_beyond-CMIP6` #190

JorgSchwinger commented:
I'll wait with my review until #190 has been merged...

jmaerz commented:
Hi @JorgSchwinger and @TomasTorsvik , I briefly discussed with @TomasTorsvik about the history issue and tested out, whether an initial merge with the `feature-hamocc_beyond-CMIP6` branch would solve the history issue (**it does**). Given our common agreement on still merging features into master that do not break backward compatibility, I am now moving the pull request to the `master` branch. I will check again with regards to the loop order (thanks @JorgSchwinger).

jmaerz committed:
changed loop order in touched loops, where possible

TomasTorsvik commented on hamocc/mo_sedmnt.F90:
```diff
@@ -70,13 +70,19 @@ MODULE mo_sedmnt
-      REAL, save :: porwat(ks)  = 0.0
+      !REAL, save :: porsol(ks)  = 0.0
+      !REAL, save :: porwah(ks)  = 0.0
+      !REAL, save :: porwat(ks)  = 0.0
```
Maybe these lines should be deleted instead of commented out? With the modified code theses lines can not be re-enabled by un-commenting.

jmaerz replied:
Right. I missed to clean that. I'll do this.

jmaerz replied:
Done.

TomasTorsvik commented on hamocc/bodensed.F90:
```diff
@@ -79,33 +82,60 @@ subroutine bodensed(kpie,kpje,kpke,pddpo)
+  ! FOR BACKWARDS COMPATIBILITY
+  !disso=3.e-8  ! (2011-01-04) EMR
+  !disso=1.e-6 ! test vom 03.03.04 half live sil ca. 20.000 yr 
+  disso_sil = 1.e-6*dtbgc
```
```suggestion
  ! Dissolution rate constant of opal (disso) [1/(kmol Si(OH)4/m3)*1/sec]
  ! THIS NEEDS TO BE CHANGED TO disso=3.e-8! THIS IS ONLY KEPT FOR THE MOMENT
  ! FOR BACKWARDS COMPATIBILITY
  !disso_sil = 3.e-8*dtbgc  ! (2011-01-04) EMR
  !disso_sil = 1.e-6*dtbgc  ! test vom 03.03.04 half live sil ca. 20.000 yr 
  disso_sil = 1.e-6*dtbgc
```
Maybe a bit pedantic, but I think it is best if the commented lines also reflect the code change.

jmaerz replied:
Correct, I can do this.

jmaerz replied:
Done.

TomasTorsvik commented on hamocc/bodensed.F90:
```diff
@@ -79,33 +82,60 @@ subroutine bodensed(kpie,kpje,kpke,pddpo)
+     enddo
   enddo
+  
+  sedict = 1.e-9 * dtbgc ! Moecular diffusion coefficient
```
```suggestion
  sedict = 1.e-9 * dtbgc ! Molecular diffusion coefficient
```
typo

jmaerz replied:
Done.

TomasTorsvik commented on hamocc/bodensed.F90:
```diff
@@ -79,33 +82,60 @@ subroutine bodensed(kpie,kpje,kpke,pddpo)
+  silsat    = 0.001
+
+  ! Degradation rate constant of POP (disso) [1/(kmol O2/m3)*1/sec]
+  disso_poc = 0.01 / 86400. * dtbgc  !  disso=3.e-5 was quite high
```
```suggestion
  disso_poc = 0.01 / 86400. * dtbgc
```
Is this comment relevant, do we need to keep it around?

jmaerz replied:
I can't tell - maybe @JorgSchwinger made some experiences that allow for better judgement?

TomasTorsvik commented on hamocc/bodensed.F90:
```diff
@@ -79,33 +82,60 @@ subroutine bodensed(kpie,kpje,kpke,pddpo)
 
-  sedict = 1.e-9 * dtbgc
+  ! Denitrification rate constant of POP (disso) [1/sec]
+  sed_denit =  0.01/86400. * dtbgc !ik      denit = 1.e-6*dtbgc
```
```suggestion
  sed_denit =  0.01/86400. * dtbgc
```
Same as above

jmaerz replied:
@JorgSchwinger - any reason to keep the comment?

JorgSchwinger replied:
No, I don't think so, too.

jmaerz replied:
Removed.

TomasTorsvik commented on hamocc/sedshi.F90:
```diff
@@ -219,7 +219,7 @@ SUBROUTINE SEDSHI(kpie,kpje,omask)
 ! fill the last active layer
         refill=seddef/(buried+1.e-10) 
-        frac = porsol(ks)*seddw(ks) !changed k to ks, ik
+        frac = porsol(i,j,ks)*seddw(ks) !changed k to ks, ik
```
```suggestion
        frac = porsol(i,j,ks)*seddw(ks)
```
another comment that doesn't seem to add much info

jmaerz replied:
I can delete it, sure.

TomasTorsvik reviewed (approved):
This looks fine to me. I wonder if we should use the opportunity to clean up some of the comments that seem outdated.

jmaerz committed:
Minor changes to meet some review comments

JorgSchwinger commented on cime_config/buildnml:
```diff
@@ -237,6 +237,9 @@ set OALKSCEN = "''"
 set OALKFILE = "''"
 set WITH_DMSPH = .false.
 set PI_PH_FILE = "''"
+set L_SED_POR = .false.
```
Maybe a more suggestive name would be preferable? For example L_SEDPOR_2D to indicate that this activates a 2d-spatially variable sediment porosity?

jmaerz replied:
I agree. Thanks for the suggestion! Maybe `l_3Dvarsedpor` (variable sediment porosity)?!

jmaerz replied:
Done

JorgSchwinger commented on cime_config/buildnml:
```diff
@@ -1470,6 +1479,8 @@ cat >>! $RUNDIR/ocn_in$inststr << EOF
+  L_SED_POR    = $L_SED_POR
+  SEDPORFILE   = $SEDPORFILE
 /
 
```
GitHub won't let me add a comment in an untouched part of the code, but further below at line 1840 there is a file `blom.input_data_list` created. If L_SED_POR is activated and SEDPORFILE specified, the filename should be added to this file (used by the CESM scripting system to check at submission whether all files are available)

jmaerz replied:
I see. Thanks for the explanation! I'll add that.

jmaerz replied:
Done

JorgSchwinger commented on hamocc/bodensed.F90:
```diff
@@ -79,33 +82,60 @@ subroutine bodensed(kpie,kpje,kpke,pddpo)
+  ! this initialization can be done later via reading a porosity map
+  if (l_sed_por)then
+   ! lon-lat variable sediment porosity from input file
+   call apply_sedpor(kpie,kpje,ks,omask,sed_por)
```
I find it a bit overkill to define an extra module for ~10 lines of code (essentially porwat=sedpor) that could be just placed directly here.

jmaerz replied:
:-) - agree - I was debating with myself - I'll change it.

JorgSchwinger commented on hamocc/mo_read_sedpor.F90:
```diff
@@ -0,0 +1,130 @@
+
+contains
+
+subroutine ini_read_sedpor(kpie,kpje,ks,omask)
```
Actually, there is no need to allocate the array sed_por permanently: The array is only needed to transfer the data from this routine to bodensed (both called from hamocc_ini). So one could define sed_por as a local (automatic) array in hamocc_ini, pass it first to read_sedpor and then to bodensed. On return from hamocc_ini, sed_por will be automatically deallocated. I would then also rename this routine to `read_sedpor` because then there is nothing that is really initialised (this would be more consistent with the fact that this is a one time read operation, not something that is used later on).

On a side note, one could integrate `bodensed` into `mo_sedmnt` and rename it to `ini_sedmnt`, which would make more sense than having it as a stand-alone subroutine. Maybe something for later.

jmaerz replied:
Hi @JorgSchwinger , I am doing the first paragraph. With respect to the second part: I like the idea, was shortly about to do it already before. I would like to do this as another pull request, though, to minimize conflicts with my own developments. I'll move the subroutine soon.

jmaerz committed:
Response to reviews

rename switch, application directly in bodensed, read porosity file to non-permanent sed_por

jmaerz committed:
merge with master

jmaerz committed:
removed comment in sedshi.F90

jmaerz commented:
Hi @JorgSchwinger and @TomasTorsvik , I adjusted the code wrt to Jörgs comments and merged with current `master` to resolve upcoming conflicts (due to previous changes to the `powtra -> ocetra` mapping). The only thing I haven't changed thus far is the comment from Tomas: https://github.com/NorESMhub/BLOM/pull/189#discussion_r962732336 which just concerns a comment in the code. Do you want to re-review the pull request or shall I merge into `master`?

TomasTorsvik commented:
> The only thing I haven't changed thus far is the comment from Tomas: [#189 (comment)](https://github.com/NorESMhub/BLOM/pull/189#discussion_r962732336) which just concerns a comment in the code.

The comment is not essential. From my point of view you can merge it as it is into `master`.

JorgSchwinger commented:
Yes, I agree, please go ahead
